nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -1377,41 +1362,80 @@ impl DataStore {
                 )),
             );
 
-        let _instance = stmt
-            .detach_and_get_result_async(
-                &*self.pool_connection_authorized(opctx).await?,
-            )
-            .await
-            .map_err(|e| match e {
-                DetachManyError::CollectionNotFound => Error::not_found_by_id(
-                    ResourceType::Instance,
-                    &authz_instance.id(),
-                ),
-                DetachManyError::NoUpdate { collection } => {
-                    if collection.runtime_state.propolis_id.is_some() {
-                        return Error::invalid_request(
+        stmt.detach_and_get_result_async(
+            &*self.pool_connection_authorized(opctx).await?,
+        )
+        .await
+        .map(|_instance| ())
+        .or_else(|e| match e {
+            // Note that if the instance is not found, we return "Ok" explicitly here.
+            //
+            // This is important for idempotency - if we crashed after setting the state,
+            // but before doing cleanup, we allow other cleanup to make progress.
+            //
+            // See also: "test_instance_deletion_is_idempotent".
+            DetachManyError::CollectionNotFound => Ok(()),
```

Comment:
Oh, this is a nicer way to handle this than the way I would've done it: I would have just added `Destroyed` to the list of "ok to delete instance states". But, I think your version is nicer, since the collection member detachment occurs in the same atomic query as the state transition, so we don't need to do it again if the instance is `Destroyed` already.

Reply:
Yeah, and also, it doesn't cause "time_deleted" to get set again